pkg/controller/control.go

```diff
@@ -1244,29 +1248,153 @@ func (c *Controller) handleErrorNoLock(err error) error {
+				log.Infof("Setting RW replica %s to ERR due to no space", address)
+				c.setReplicaModeNoLock(address, types.ERR)
+			}
+		}
+		return true
 	}
-	for address := range replicaNoSpaceErrMap {
+
+	for address := range rwReplicasOnEnospcMap {
+		log.Infof("Setting RW replica %s to ERR due to no space", address)
 		c.setReplicaModeNoLock(address, types.ERR)
 	}
-	return nil
+
+	return false
+}
+
+// categorizeOutOfSpaceReplicas returns the count of all RW replicas, a map of RW replicas on no space disks, and a list of WO replicas on no space disks.
+// The functions separate RW/WO replicas in replicaWrittenBytesMap into a RW replicas with WriteBytes Map and a WO replicas list to be set to `ERR`.
+//
+//	replicaWrittenBytesMap: a map of replicas that encountered no space error and the bytes they have written when encountering the error. For example: {"replica A": 1, "replica B": 2, "replica C": 2, "replica D": 2, "replica E": 0,}
+func (c *Controller) categorizeOutOfSpaceReplicas(replicaWrittenBytesMap map[string]int) (int, map[string]int, []string) {
+	rwReplicaCount := 0
+	noSpaceRWReplicasMap := make(map[string]int, len(c.replicas))
+	noSpaceWOReplicasList := make([]string, 0, len(c.replicas))
+
+	for _, r := range c.replicas {
+		switch r.Mode {
+		case types.RW:
+			rwReplicaCount++
+			// The RW replicas that are not on out of space disks will not in the replicaWrittenBytesMap.
+			if wbn, exist := replicaWrittenBytesMap[r.Address]; exist {
+				noSpaceRWReplicasMap[r.Address] = wbn
+			}
+		case types.WO:
+			if _, exist := replicaWrittenBytesMap[r.Address]; exist {
+				noSpaceWOReplicasList = append(noSpaceWOReplicasList, r.Address)
+			}
+		}
+	}
+
+	return rwReplicaCount, noSpaceRWReplicasMap, noSpaceWOReplicasList
+}
+
+// listReplicasToErrOnEnospc lists the replicas that need to be set to ERR on ENOSPC error,
+// and retain max number of replicas that have the same write bytes to keep the volume working.
+//
+//	return a list of replicas that need to be set to ERR on ENOSPC error.
+//	For example:
+//	replicaToErrOnNoSpaceMap are {"replica A": 1, "replica B": 2, "replica C": 1, "replica D": 2, "replica E": 0,}
+//	Written Bytes 2: {"replica B", "replica D"},
+//	Written Bytes 1: {"replica A", "replica C"},
+//	Written Bytes 0: {"replica E"}
+//	{"replica B", "replica D"} will be retained RW because they have max number of the replicas that wrote the same written bytes, and the written bytes is maximum.
+//	Therefore, it will return the list {"replica A", "replica C", "replica E"} to be set to ERR.
+func listReplicasToErrOnEnospc(replicaToErrOnNoSpaceMap map[string]int) []string {
+	writeBytesReplicasMap := getReplicaWriteBytesMap(replicaToErrOnNoSpaceMap)
+	maxWrittenBytes := getMaxWrittenBytesForMaxNumberReplicas(writeBytesReplicasMap)
+
+	// Retain replicas in writeBytesReplicasMap[maxWrittenBytes] RW
+	return getFilteredReplicasToErrList(replicaToErrOnNoSpaceMap, writeBytesReplicasMap[maxWrittenBytes])
```

Comment:
```suggestion
	return listReplicasToErr(replicaToErrOnNoSpaceMap, retainedReplicas)

```

Reply:
Removed func `getFilteredReplicasToErrList` and moved the logic to func `listReplicasToErrOnEnospc`